### PR TITLE
Make two SubReaderWrapper implementations singletons

### PR DESCRIPTION
**Pull request by original-brownbear**
random find:

No state in either of these, no need to recreate them for every directory. Saves a few cycles and makes it more obvious that there's no state in these.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/lucene/Lucene.java [hunk 1]
```diff
@@ -884,24 +884,26 @@ protected StoredFieldsReader doGetSequentialStoredFieldsReader(StoredFieldsReade
...
-                        return new LeafReaderWithLiveDocs(leaf, null, leaf.maxDoc());
-                    }
-                    // Once soft-deletes is enabled, we no longer hard-update or hard-delete documents directly.
-                    // Two scenarios that we have hard-deletes: (1) from old segments where soft-deletes was disabled,
-                    // (2) when IndexWriter hits non-aborted exceptions. These two cases, IW flushes SegmentInfos
-                    // before exposing the hard-deletes, thus we can use the hard-delete count of SegmentInfos.
-                    final int numDocs = segmentReader.maxDoc() - segmentReader.getSegmentInfo().getDelCount();
-                    assert numDocs == popCount(hardLiveDocs) : numDocs + " != " + popCount(hardLiveDocs);
-                    return new LeafReaderWithLiveDocs(segmentReader, hardLiveDocs, numDocs);
+        private static final SubReaderWrapper SUB_READER_WRAPPER = new SubReaderWrapper() {
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/common/lucene/Lucene.java:887]:
```suggestion
        private static final SubReaderWrapper ALL_LIVE_DOCS_SUB_READER_WRAPPER = new SubReaderWrapper() {
```

**Review by tlrx — APPROVED** [review]
LGTM, left a minor renaming suggestion

**Comment by original-brownbear** [comment]
Thanks Tanguy!

**Comment by original-brownbear** [comment]
@elasticmachine update branch

**Pull request merged**